pkg/scheduler/controller/backupbucket/scheduler_reconciler.go

```diff
@@ -0,0 +1,229 @@
+	gardenhelper "github.com/gardener/gardener/pkg/apis/garden/v1beta1/helper"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// MsgUnschedulable is the Message for the Event on a BackupBucket that the Scheduler creates in case it cannot schedule the BackupBucket to any Seed
+const MsgUnschedulable = "Failed to schedule backupbucket"
+
+func (c *SchedulerController) backupBucketAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	newBackupBucket := obj.(*gardencorev1alpha1.BackupBucket)
+
+	if newBackupBucket.DeletionTimestamp != nil {
+		logger.Logger.Infof("Ignoring backupBucket '%s' because it has been marked for deletion", newBackupBucket.Name)
+		c.backupBucketQueue.Forget(key)
+		return
+	}
+
+	c.backupBucketQueue.Add(key)
+}
+
+func (c *SchedulerController) backupBucketUpdate(oldObj, newObj interface{}) {
+	c.backupBucketAdd(newObj)
+}
+
+// reconciler implements the reconcile.Reconcile interface for backupBucket scheduler.
+type reconciler struct {
+	ctx      context.Context
+	client   client.Client
+	recorder record.EventRecorder
+	logger   *logrus.Entry
+}
+
+// newReconciler returns the new backupBucker reconciler.
+func newReconciler(ctx context.Context, gardenClient client.Client, recorder record.EventRecorder) reconcile.Reconciler {
+	return &reconciler{
+		ctx:      ctx,
+		client:   gardenClient,
+		recorder: recorder,
+		logger:   logger.NewFieldLogger(logger.Logger, "scheduler", "backupbucket"),
+	}
+}
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	bb := &gardencorev1alpha1.BackupBucket{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, bb); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("[SCHEDULER BACKUPBUCKET RECONCILE] %s - skipping because BackupBucket has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Infof("[SCHEDULER BACKUPBUCKET RECONCILE] %s - unable to retrieve object from store: %v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, r.scheduleBackupBucket(bb)
+}
+
+type executeSchedulingRequest = func(context.Context, *gardencorev1alpha1.BackupBucket) error
+
+func (r *reconciler) scheduleBackupBucket(obj *gardencorev1alpha1.BackupBucket) error {
+	var (
+		backupBucket    = obj.DeepCopy()
+		schedulerLogger = r.logger.WithField("backupbucket", backupBucket.Name)
+	)
+
+	if backupBucket.Spec.Seed != nil {
+		return nil
+		// If the BackupBucket manifest already specifies a desired Seed cluster,
+		// we should check its availability. If its not available we will try reschedule it again.
```

Comment:
This can only work if the backup bucket was not yet created, right? Or how would that look like for existing backup buckets with lots of entries inside if the seed is unavailable for a few minutes?

Reply:
Nope. From core.backupBuckets perspective seed is just a worker for reconcilation by rdeploying extensions.backupBucekt resource in that seed. If seed is down, the extension.backupBucket  resource will be deployed on new seed, which will make sure bucket is there. The existing entries will still be there in bucket. Since bucket spec i.e. bucketName, secret, provider and region still remains same as `core.BackupBucket` is still same.  I think you are in for it, so I'll uncomment the commented part.

Reply:
>  If seed is down, the extension.backupBucket resource will be deployed on new seed,

What happens then to the `extension.BackupBucket` resource on the old seed?

Reply:
Sorry for late reply. 
Here is how i'm planning to implementing it,
- In case of actual disaster where seed is not reachable we cannot do much right now. This is more off control-plane migration / seed disaster recovery topic.  
- In case its reachable, i'll simply remove "Gardener" finalizer and then apply deletionTimestamp, since we don't want deletion of resource to delete actual bucket.